From a080cfd763e72e221b3769efc47315749daa4869 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Mon, 31 Mar 2025 23:42:02 +0200
Subject: [PATCH 1668/1695] drivers: stm32_risaf: refactor configuration
 functions

Factorize source code around region and subregion configuration
functions by passing structure pointers instead of unitary
variables for each parameter. Prefer to extract them directly
within the functions.
Implement a save/restore mechanism in stm32_risaf_reconfigure_area()
in case of error to keep consistency.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0ed8c80ad8c738507e4ae00b57500edd80dffa42 drivers: stm32_risaf: refactor configuration functions]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: Ied0674c69593aa8efdec9b6c436f0e85e8376a85
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/447776
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/firewall/stm32_risaf.c | 165 +++++++++++-----------------
 1 file changed, 62 insertions(+), 103 deletions(-)

diff --git a/core/drivers/firewall/stm32_risaf.c b/core/drivers/firewall/stm32_risaf.c
index ca3290d4c..110edb9fc 100644
--- a/core/drivers/firewall/stm32_risaf.c
+++ b/core/drivers/firewall/stm32_risaf.c
@@ -492,13 +492,20 @@ risaf_check_overlap(struct stm32_risaf_instance *risaf __maybe_unused,
 	return TEE_SUCCESS;
 }
 
-static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
-					 uint32_t region_id, uint32_t cfg,
-					 uint32_t cid_cfg, paddr_t saddr,
-					 paddr_t eaddr)
+static TEE_Result
+risaf_configure_region(struct stm32_risaf_instance *risaf,
+		       const struct stm32_risaf_region *region)
 {
-	uint32_t mask = risaf->ddata->mask_regions;
 	vaddr_t base = risaf_base(risaf);
+	paddr_t start_addr = region->addr;
+	paddr_t end_addr = start_addr + region->len - 1U;
+	uint32_t id = _RISAF_GET_REGION_ID(region->cfg);
+	uint32_t cfg = stm32_risaf_get_region_config(region->cfg);
+	uint32_t cid_cfg = stm32_risaf_get_region_cid_config(region->cfg);
+	uint32_t mask = risaf->ddata->mask_regions;
+
+	DMSG("Reconfiguring %s region ID: %"PRIu32, risaf->pdata.risaf_name,
+	     id);
 
 	if (cfg & _RISAF_REG_CFGR_ENC) {
 		if (!risaf->pdata.enc_supported) {
@@ -525,41 +532,49 @@ static TEE_Result risaf_configure_region(struct stm32_risaf_instance *risaf,
 		}
 	}
 
-	io_clrbits32(base + _RISAF_REG_CFGR(region_id), _RISAF_REG_CFGR_BREN);
+	io_clrbits32(base + _RISAF_REG_CFGR(id), _RISAF_REG_CFGR_BREN);
 
-	io_clrsetbits32(base + _RISAF_REG_STARTR(region_id), mask,
-			(saddr - risaf->pdata.mem_base) & mask);
-	io_clrsetbits32(base + _RISAF_REG_ENDR(region_id), mask,
-			(eaddr - risaf->pdata.mem_base) & mask);
-	io_clrsetbits32(base + _RISAF_REG_CIDCFGR(region_id),
+	io_clrsetbits32(base + _RISAF_REG_STARTR(id), mask,
+			(start_addr - risaf->pdata.mem_base) & mask);
+	io_clrsetbits32(base + _RISAF_REG_ENDR(id), mask,
+			(end_addr - risaf->pdata.mem_base) & mask);
+	io_clrsetbits32(base + _RISAF_REG_CIDCFGR(id),
 			_RISAF_REG_CIDCFGR_ALL_MASK,
 			cid_cfg & _RISAF_REG_CIDCFGR_ALL_MASK);
 
-	io_clrsetbits32(base + _RISAF_REG_CFGR(region_id),
-			_RISAF_REG_CFGR_ALL_MASK,
+	io_clrsetbits32(base + _RISAF_REG_CFGR(id), _RISAF_REG_CFGR_ALL_MASK,
 			cfg & _RISAF_REG_CFGR_ALL_MASK);
 
 	DMSG("RISAF %#"PRIxPA": region %02"PRIu32" - start %#"PRIxPA
 	     "- end %#"PRIxPA" - cfg %#08"PRIx32" - cidcfg %#08"PRIx32,
-	     risaf->pdata.base.pa, region_id,
-	     risaf->pdata.mem_base +
-	     io_read32(base + _RISAF_REG_STARTR(region_id)),
-	     risaf->pdata.mem_base +
-	     io_read32(base + _RISAF_REG_ENDR(region_id)),
-	     io_read32(base + _RISAF_REG_CFGR(region_id)),
-	     io_read32(base + _RISAF_REG_CIDCFGR(region_id)));
+	     risaf->pdata.base.pa, id,
+	     risaf->pdata.mem_base + io_read32(base + _RISAF_REG_STARTR(id)),
+	     risaf->pdata.mem_base + io_read32(base + _RISAF_REG_ENDR(id)),
+	     io_read32(base + _RISAF_REG_CFGR(id)),
+	     io_read32(base + _RISAF_REG_CIDCFGR(id)));
 
 	return TEE_SUCCESS;
 }
 
-static TEE_Result risaf_configure_subregion(struct stm32_risaf_instance *risaf,
-					    uint32_t reg_id,
-					    uint32_t subreg_id, uint32_t cfg,
-					    uint32_t nest_cfg, uintptr_t saddr,
-					    uintptr_t eaddr)
+static TEE_Result
+risaf_configure_subregion(struct stm32_risaf_instance *risaf,
+			  uint32_t reg_id,
+			  const struct stm32_risaf_subregion *subregion)
 {
 	vaddr_t base = risaf_base(risaf);
+	paddr_t start_addr = subregion->addr;
+	paddr_t end_addr = start_addr + subregion->len - 1U;
 	uint32_t mask = risaf->ddata->mask_regions;
+	uint32_t subreg_id = _RISAF_GET_SUBREGION_ID(subregion->cfg);
+	uint32_t cfg = stm32_risaf_get_subregion_config(subregion->cfg);
+	uint32_t nest_cfg =
+		stm32_risaf_get_subregion_nest_config(subregion->cfg);
+
+	assert(subreg_id < risaf->ddata->max_subregions);
+
+	DMSG("Configuring %s subregion ID: %"PRIu32" region ID: %"PRIu32,
+	     risaf->pdata.risaf_name, _RISAF_GET_SUBREGION_ID(subregion->cfg),
+	     reg_id);
 
 	if (cfg & _RISAF_SUBREG_CFGR_RLOCK) {
 		EMSG("RISAF %#"PRIxPA": can't configure locked subregion",
@@ -571,9 +586,9 @@ static TEE_Result risaf_configure_subregion(struct stm32_risaf_instance *risaf,
 		     _RISAF_SUBREG_CFGR_SREN);
 
 	io_clrsetbits32(base + _RISAF_SUBREG_STARTR(reg_id, subreg_id), mask,
-			(saddr - risaf->pdata.mem_base) & mask);
+			(start_addr - risaf->pdata.mem_base) & mask);
 	io_clrsetbits32(base + _RISAF_SUBREG_ENDR(reg_id, subreg_id), mask,
-			(eaddr - risaf->pdata.mem_base) & mask);
+			(end_addr - risaf->pdata.mem_base) & mask);
 	io_clrsetbits32(base + _RISAF_SUBREG_NESTR(reg_id, subreg_id),
 			_RISAF_SUBREG_NESTR_ALL_MASK,
 			nest_cfg & _RISAF_SUBREG_NESTR_ALL_MASK);
@@ -669,39 +684,18 @@ static TEE_Result stm32_risaf_pm_resume(struct stm32_risaf_instance *risaf)
 
 	for (i = 0; i < risaf->pdata.nregions; i++) {
 		uint32_t id = _RISAF_GET_REGION_ID(regions[i].cfg);
-		paddr_t start_addr = 0;
-		paddr_t end_addr = 0;
-		uint32_t cid_cfg = 0;
-		uint32_t cfg = 0;
 		unsigned int j = 0;
 		struct stm32_risaf_subregion *subreg = regions[i].subregions;
 
 		if (!id)
 			continue;
 
-		cfg = stm32_risaf_get_region_config(regions[i].cfg);
-		cid_cfg = stm32_risaf_get_region_cid_config(regions[i].cfg);
-		start_addr = regions[i].addr;
-		end_addr = start_addr + regions[i].len - 1U;
-		if (risaf_configure_region(risaf, id, cfg, cid_cfg,
-					   start_addr, end_addr))
+		if (risaf_configure_region(risaf, &regions[i]))
 			panic();
 
 		for (j = 0; j < regions[i].nsubregions; j++) {
-			uint32_t subreg_id =
-				_RISAF_GET_SUBREGION_ID(subreg[j].cfg);
-			uint32_t nest_cfg = 0;
-
-			cfg = stm32_risaf_get_subregion_config(subreg[j].cfg);
-			nest_cfg =
-			stm32_risaf_get_subregion_nest_config(subreg[j].cfg);
 
-			start_addr = subreg[j].addr;
-			end_addr = start_addr + subreg[j].len - 1U;
-
-			if (risaf_configure_subregion(risaf, id, subreg_id,
-						      cfg, nest_cfg,
-						      start_addr, end_addr))
+			if (risaf_configure_subregion(risaf, id, &subreg[j]))
 				panic();
 		}
 	}
@@ -1030,34 +1024,26 @@ static TEE_Result stm32_risaf_reconfigure_area(struct firewall_query *fw,
 	exceptions = cpu_spin_lock_xsave(&risaf->pdata.conf_lock);
 
 	if (is_region) {
-		DMSG("Reconfiguring %s region ID: %"PRIu32,
-		     risaf->pdata.risaf_name, id);
-		res =
-		risaf_configure_region(risaf, id,
-				       stm32_risaf_get_region_config(q_cfg),
-				       stm32_risaf_get_region_cid_config(q_cfg),
-				       region->addr,
-				       region->addr + region->len - 1);
+		uint32_t cfg_save = region->cfg;
+
+		region->cfg = q_cfg;
+
+		res = risaf_configure_region(risaf, region);
+		/* Restore initial value if configuration fails */
+		if (res)
+			region->cfg = cfg_save;
 	} else {
 		struct stm32_risaf_subregion *subreg = NULL;
-		uint32_t subreg_id = 0;
-		uint32_t cfg = 0;
-		uint32_t nest = 0;
+		uint32_t cfg_save = 0;
 
 		subreg = &region->subregions[subregion_idx];
-		subreg_id = _RISAF_GET_SUBREGION_ID(subreg->cfg);
-
-		DMSG("Reconfiguring %s subregion ID: %"PRIu32
-		     " region ID: %"PRIu32,
-		     risaf->pdata.risaf_name, subreg_id, id);
-
-		cfg = stm32_risaf_get_subregion_config(q_cfg);
-		nest = stm32_risaf_get_subregion_nest_config(q_cfg);
+		cfg_save = subreg->cfg;
+		subreg->cfg = q_cfg;
 
-		res =
-		risaf_configure_subregion(risaf, id, subreg_id, cfg, nest,
-					  subreg->addr,
-					  subreg->addr + subreg->len - 1);
+		res = risaf_configure_subregion(risaf, id, subreg);
+		/* Restore initial value if configuration fails */
+		if (res)
+			subreg->cfg = cfg_save;
 	}
 
 	cpu_spin_unlock_xrestore(&risaf->pdata.conf_lock, exceptions);
@@ -1156,11 +1142,7 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 	for (i = 0; i < nregions; i++) {
 		const fdt32_t *prop = NULL;
 		const fdt32_t *subconf_list = NULL;
-		paddr_t start_addr = 0;
-		paddr_t end_addr = 0;
-		uint32_t cid_cfg = 0;
 		uint32_t phandle = 0;
-		uint32_t cfg = 0;
 		uint32_t id = 0;
 		int pnode = 0;
 		unsigned int j = 0;
@@ -1207,15 +1189,7 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 		id = _RISAF_GET_REGION_ID(regions[i].cfg);
 		assert(id < risaf->ddata->max_base_regions);
 
-		cfg = stm32_risaf_get_region_config(regions[i].cfg);
-
-		cid_cfg = stm32_risaf_get_region_cid_config(regions[i].cfg);
-
-		start_addr = regions[i].addr;
-		end_addr = start_addr + regions[i].len - 1U;
-
-		if (risaf_configure_region(risaf, id, cfg, cid_cfg,
-					   start_addr, end_addr))
+		if (risaf_configure_region(risaf, &regions[i]))
 			panic();
 
 		/*  Consider subregions if any */
@@ -1228,7 +1202,7 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 		if (regions[i].nsubregions > risaf->ddata->max_subregions)
 			panic();
 
-		regions[i].subregions =	calloc(regions[i].nsubregions,
+		regions[i].subregions = calloc(regions[i].nsubregions,
 					       sizeof(*regions[i].subregions));
 		if (regions[i].nsubregions && !regions[i].subregions) {
 			EMSG("Out of memory in node %s",
@@ -1240,8 +1214,6 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 
 		for (j = 0; j < regions[i].nsubregions; j++) {
 			int subnode = 0;
-			uint32_t subreg_id = 0;
-			uint32_t nest_cfg = 0;
 
 			phandle = fdt32_to_cpu(*(subconf_list + j));
 			subnode = fdt_node_offset_by_phandle(fdt, phandle);
@@ -1272,20 +1244,7 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 							     &regions[i], j))
 				panic();
 
-			subreg_id = _RISAF_GET_SUBREGION_ID(subreg[j].cfg);
-			assert(subreg_id < risaf->ddata->max_subregions);
-
-			cfg = stm32_risaf_get_subregion_config(subreg[j].cfg);
-
-			nest_cfg =
-			stm32_risaf_get_subregion_nest_config(subreg[j].cfg);
-
-			start_addr = subreg[j].addr;
-			end_addr = start_addr + subreg[j].len - 1U;
-
-			if (risaf_configure_subregion(risaf, id, subreg_id,
-						      cfg, nest_cfg,
-						      start_addr, end_addr))
+			if (risaf_configure_subregion(risaf, id, &subreg[j]))
 				panic();
 		}
 	}
-- 
2.47.2

